File: src/domain/label/services/LabelAggregator.js
/* exported LabelAggregator */

class LabelAggregator {
  /**
   * BaseRow[] から SKU 数量を集計して LabelItem[] を返す
   * @param {any[]} rows
   * @returns {LabelItem[]}
   */
  aggregate(rows) {
    const map = new Map();

    for (const row of rows || []) {
      const sku = String(row.get("SKU") || '').trim();
      const quantity = Number(row.get("購入数") || 0) || 0;

      if (!sku || quantity <= 0) continue;
      map.set(sku, (map.get(sku) || 0) + quantity);
    }

    const items = Array.from(map.entries()).map(([sku, qty]) => new LabelItem(sku, qty));
    if (items.length === 0) {
      throw new Error('有効なSKUがありません');
    }
    return items;
  }
}

File: python/src/domain/label/services/label_aggregator.py
from __future__ import annotations
from typing import Protocol
from domain.label.entities.label_item import LabelItem


class RowLike(Protocol):
    def get(self, column_name: str) -> str: ...


class LabelAggregator:
    def aggregate(self, rows: list[RowLike]) -> list[LabelItem]:
        sku_totals: dict[str, int] = {}
        for row in rows or []:
            sku = str(row.get("SKU") or "").strip()
            raw_qty = row.get("購入数")
            quantity = int(raw_qty) if raw_qty else 0
            if not sku or quantity <= 0:
                continue
            sku_totals[sku] = sku_totals.get(sku, 0) + quantity
        items = [LabelItem(sku=sku, quantity=qty) for sku, qty in sku_totals.items()]
        if not items:
            raise ValueError("有効なSKUがありません")
        return items
